third_party/rust/fpe/v0_6/crate/src/ff1.rs

```diff
@@ -0,0 +1,469 @@
+
+mod error;
+pub use error::{InvalidRadix, NumeralStringError};
+
+#[cfg(feature = "alloc")]
+mod alloc;
+#[cfg(feature = "alloc")]
+pub use self::alloc::{BinaryNumeralString, FlexibleNumeralString};
+
+#[cfg(test)]
+mod proptests;
+
+#[cfg(test)]
+mod test_vectors;
+
+/// The minimum allowed numeral string length for any radix.
+const MIN_NS_LEN: u32 = 2;
+/// The maximum allowed numeral string length for any radix.
+const MAX_NS_LEN: usize = u32::MAX as usize;
+
+/// The minimum allowed value of radix^minlen.
+///
+/// Defined in [NIST SP 800-38G Revision 1](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-38Gr1-draft.pdf).
+const MIN_NS_DOMAIN_SIZE: u64 = 1_000_000;
+
+/// `minlen` such that `2^minlen >= MIN_NS_DOMAIN_SIZE`.
+const MIN_RADIX_2_NS_LEN: u32 = 20;
+
+#[cfg(test)]
+const_assert!((1 << MIN_RADIX_2_NS_LEN) >= MIN_NS_DOMAIN_SIZE);
+
+#[derive(Debug, PartialEq)]
+enum Radix {
+    /// A radix in [2..2^16]. It uses floating-point arithmetic.
+    Any { radix: u32, min_len: u32 },
+    /// A radix 2^i for i in [1..16]. It does not use floating-point arithmetic.
+    PowerTwo {
+        radix: u32,
+        min_len: u32,
+        log_radix: u8,
+    },
+}
+
+impl Radix {
+    fn from_u32(radix: u32) -> Result<Self, InvalidRadix> {
+        // radix must be in range [2..=2^16]
+        if !(2..=(1 << 16)).contains(&radix) {
+            return Err(InvalidRadix(radix));
+        }
+
+        Ok(if radix.count_ones() == 1 {
+            let log_radix = 31 - radix.leading_zeros();
+            Radix::PowerTwo {
+                radix,
+                min_len: cmp::max((MIN_RADIX_2_NS_LEN + log_radix - 1) / log_radix, MIN_NS_LEN),
+                log_radix: u8::try_from(log_radix).unwrap(),
+            }
+        } else {
+            // If radix is exactly 2^16 then the code below would overflow a u32.
+            let radix_u64 = u64::from(radix);
+            let mut min_len = 1u32;
+            let mut domain = radix_u64;
+            while domain < MIN_NS_DOMAIN_SIZE {
+                domain *= radix_u64;
+                min_len += 1;
+            }
+            Radix::Any { radix, min_len }
+        })
+    }
+
+    fn check_ns_length(&self, ns_len: usize) -> Result<(), NumeralStringError> {
+        let min_len = match *self {
+            Radix::Any { min_len, .. } => min_len as usize,
+            Radix::PowerTwo { min_len, .. } => min_len as usize,
+        };
+        let max_len = MAX_NS_LEN;
+
+        if ns_len < min_len {
+            Err(NumeralStringError::TooShort { ns_len, min_len })
+        } else if ns_len > max_len {
+            Err(NumeralStringError::TooLong { ns_len, max_len })
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Calculates b = ceil(ceil(v * log2(radix)) / 8).
+    fn calculate_b(&self, v: usize) -> usize {
+        use libm::{ceil, log2};
+        match *self {
+            Radix::Any { radix, .. } => ceil(v as f64 * log2(f64::from(radix)) / 8f64) as usize,
```

Comment:
Feels silly to add an entire dep for these math functions because the library works with `nostd` rust when we build both against `std` and link to system `libm`. We could ask them to make it optional, but I suppose the linker will cut out almost all the code.